File: cs/cslox/Interpreter/LoxInstance.cs
using System;
using cslox.Scanning;

namespace cslox.Interpreter;

class LoxInstance
{
	private readonly LoxClass klass;
	private readonly Dictionary<string, object?> fields = new Dictionary<string, object?>();

	public LoxInstance(LoxClass klass)
	{
		this.klass = klass;
	}

	public object? Get(Token name)
	{
		if (fields.ContainsKey(name.Lexeme))
		{
			return fields[name.Lexeme];
		}

		var method = klass.FindMethod(name.Lexeme);
		if (method != null)
		{
			return method;
		}

		throw new RuntimeError(name, $"Undefined property '{name.Lexeme}'.");
	}

	public void Set(Token name, object? value)
	{
		fields[name.Lexeme] = value;
	}

    public override string ToString()
    {
		return $"{klass.Name} instance";
    }
}

File: tree-walk-interpreter/go/internal/lox/lox_instance.go
package lox

import "fmt"

type loxInstance struct {
	class  *loxClass
	fields map[string]interface{}
}

func newLoxInstance(class *loxClass) *loxInstance {
	return &loxInstance{
		class:  class,
		fields: make(map[string]interface{}),
	}
}

func (i *loxInstance) get(name *Token) interface{} {
	if value, ok := i.fields[name.Lexeme]; ok {
		return value
	}

	panic(newRuntimeError(name, fmt.Sprintf("Undefined property '%s'.", name.Lexeme)))
}

func (i *loxInstance) set(name *Token, value interface{}) {
	i.fields[name.Lexeme] = value
}

func (i *loxInstance) String() string {
	return i.class.name + " instance"
}
